### PR TITLE
Bug #3162: Merge Node and NodeRGBA, fix opacity incorrect bug.

### PR DESCRIPTION
**Pull request by boyu0**
Deprecate RGBAProtocol, NodeRGBA and LayerRGBA，add cascade opacity and cascade color test

### CODE REVIEW

**Comment by minggo** [comment]
@ricardoquesada 
What's your opinion to merge NodeRGBA?

**Comment by minggo** [comment]
@boyu0 
I think you should deprecated old functions and classes, can not just delete them.

**Comment by dualface** [comment]
+1

Merge NodeRGBA to Node, remove all duplicate codes.

**Comment by boyu0** [comment]
@minggo  ok, I will add them back and duplicate them.

**Comment by ricardoquesada** [comment]
Interesting.

Some context:
`NodeRGBA` is kind of a weird node. Because it has `color` and `opacity` but you can't draw a `NodeRGBA`. So having this node is not very useful. Multiple inheritance would be better in this case... And a component-based model will solve it better way yet.

So, merging `NodeRGBA` into `Node` will simplify our code (+1 for this).
But will transfer the "weird" properties to `Node`. `Node` will have `opacity` and `color` but if you draw it nothing will happen.

What bug are we trying to fix with this PR ?
The inheritance of color and opacity ?
If so, then it seems OK to me.

But we should know that it transfer (or at least it doesn't fix) the "weird" issue of having a Node with `opacity` and `color`, but a Node can't be drawn.

When we move to a Component-base model, we need to fix this properly.

**Comment by minggo** [comment]
This pull request is to fix [this issue](http://www.cocos2d-x.org/issues/3162).

**Comment by boyu0** [comment]
@ricardoquesada 
I was trying to fix this bug: Child's opacity will not be changed when its parent's setCascadeOpacityEnabled was set to true and opacity was changed.
http://www.cocos2d-x.org/issues/2920
To fix this it would add lots of code, I found that classes inherit from RGBAProtocol( NodeRGBA, LayerRGBA, LabelBMFont and Label) have almost same code, it has so much duplicate code, It's hard to maintain. I thought NodeRGBA is a little overdesign, merge it to Node can simplify our engine, so I merge them.

**Comment by ricardoquesada** [comment]
@boyu0 
correct. I think merging it is the right thing to do to fix that particular bug.

**Comment by minggo** [comment]
@boyu0 
Could you please fix conflicts.
Then i will merge it.

File: cocos/2d/CCNode.h [hunk 1]
```diff
@@ -1477,6 +1499,14 @@ class CC_DLL Node : public Object
...
     PhysicsBody* _physicsBody;        ///< the physicsBody the node have
 #endif
+    
+    // opacity controls
+	GLubyte		_displayedOpacity;
+    GLubyte     _realOpacity;
+	Color3B	    _displayedColor;
+    Color3B     _realColor;
+	bool		_cascadeColorEnabled;
+    bool        _cascadeOpacityEnabled;
```
Comment (minggo) [inline cocos/2d/CCNode.h:1517]: Error indent.

File: cocos/2d/CCNode.h [hunk 2]
```diff
@@ -1485,46 +1515,38 @@ class CC_DLL Node : public Object
...
  All features from Node are valid, plus the following new features:
  - opacity
  - RGB colors
-
+ 
  Opacity/Color propagates into children that conform to the RGBAProtocol if cascadeOpacity/cascadeColor is enabled.
  @since v2.1
  */
-class CC_DLL NodeRGBA : public Node, public RGBAProtocol
+class CC_DLL CC_DEPRECATED_ATTRIBUTE NodeRGBA : public Node, public RGBAProtocol
```
Comment (minggo) [inline cocos/2d/CCNode.h]: Please put deprecated classes into Deprecated.h/cpp.

**Comment by minggo** [comment]
And please add test case for this issue.
Thanks.

**Pull request merged**